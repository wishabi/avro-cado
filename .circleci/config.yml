defaults: &defaults
  working_directory: ~/workspace
  docker:
      - image: circleci/node:10.14.2-alpine

version: 2
jobs:
  build:
    <<: *defaults
    steps:
      - checkout
      # see if we have the node_modules cached already
      - restore_cache:
          key: 'v1-node-module-cache-{{ .Branch }}-{{ checksum "package.json" }}'
      - run: npm i
      # save the node_modules cache
      - save_cache:
          key: 'v1-node-module-cache-{{ .Branch }}-{{ checksum "package.json" }}'
          paths:
            - ~/workspace/node_modules

      - persist_to_workspace:
          root: ~/workspace
          paths:
            - .
  test:
    <<: *defaults
    steps:
      - attach_workspace:
          at: ~/workspace
      - run:
          command: npm run test
      - store_test_results:
          path: reports/unit-tests
  lint:
    <<: *defaults
    steps:
      - attach_workspace:
          at: ~/workspace
      - run:
          command: npm run lint
workflows:
  version: 2
  build_and_test:
    jobs:
      - build
      - test:
          requires:
            - build