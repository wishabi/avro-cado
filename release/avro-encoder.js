"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("./util");
const config_1 = require("./config");
const rp = require("request-promise");
const Avro = require("avsc");
/**
 * Register the specified schema for the specified topic under the
 * specifed subject. The subject used is: <topic>-<key|value>
 *
 * @param subject - the subject under which to register the schema
 * @param schemaRegistry - the schema registry host
 * @param numRetries - the number of retries to register before throwing an error
 * @param schema - the schema to register
 *
 * @return - A Promise holding the id of the schema in the schema registry
 */
exports.registerSchema = ({ subject, schemaRegistry, numRetries, schema }) => __awaiter(this, void 0, void 0, function* () {
    // null schema so nothing to register
    if (!schema) {
        return -1;
    }
    // craft the REST call to the schema registry
    const req = {
        method: "POST",
        uri: `${schemaRegistry}/subjects/${subject}/versions`,
        headers: { Accept: config_1.ACCEPT_HEADERS },
        body: {
            schema: JSON.stringify(schema)
        },
        json: true,
        resolveWithFullResponse: true
    };
    let schemaId;
    let error = null;
    // implement retry on certain status/reason codes
    for (let i = 0; i <= numRetries; i += 1) {
        try {
            schemaId = (yield rp(req)).body.id;
            // make sure to clear any previous errors
            error = null;
            break;
        }
        catch (err) {
            // save the error
            error = err;
            if (util_1.handleError(err) && i + 1 <= numRetries) {
                // try and try again until we run out of retries
                continue;
            }
            // fatal error
            break;
        }
    }
    if (error) {
        throw new Error(`Failed to register schema for subject ${subject} :: ${error.message}`);
    }
    return schemaId;
});
/**
 * @todo Fix doc comment
 * Encode an Avro value into a message, as expected by Confluent's Kafka Avro
 * deserializer.
 *
 * @param message - message to encode
 */
exports.encodeData = (payload, schemaId, schema, length = 1024) => {
    if (payload === null) {
        // no payload so return it
        return null;
    }
    if (schemaId === -1) {
        // Bufferify the payload
        return Buffer.from(JSON.stringify(payload));
    }
    const buf = new Buffer(length);
    buf[0] = 0; // magic byte
    buf.writeInt32BE(schemaId, 1);
    const pos = schema.encode(payload, buf, 5);
    if (pos < 0) {
        // the buffer was too short, we need to resize
        return exports.encodeData(payload, schemaId, schema, length - pos);
    }
    return buf.slice(0, pos);
};
const genMessageEncoder = ({ schema, schemaId }) => (message) => {
    return exports.encodeData(message, schemaId, schema);
};
const createAvroSchema = ({ wrapUnions, schema }) => {
    if (schema) {
        return Avro.Type.forSchema(schema, { wrapUnions });
    }
    return null;
};
/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/
exports.createEncoder = (opts) => __awaiter(this, void 0, void 0, function* () {
    // Aggregare the configuration values with defaults
    const mergedOpts = config_1.processOptions(opts);
    // encoder info
    const encoderInfo = {
        schema: null,
        schemaId: -1
    };
    // register the value and key schemas and parse them
    [encoderInfo.schema, encoderInfo.schemaId] = yield Promise.all([
        createAvroSchema(mergedOpts),
        exports.registerSchema(mergedOpts)
    ]);
    return genMessageEncoder(encoderInfo);
});
//# sourceMappingURL=avro-encoder.js.map