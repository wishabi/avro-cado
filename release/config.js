"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util_1 = require("../ts/util");
exports.ACCEPT_HEADERS = "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";
const optionsDefault = {
    schemaRegistry: "http://localhost:8081",
    numRetries: 10,
    wrapUnions: "auto"
};
exports.processOptions = opts => {
    if (!opts.subject) {
        throw new Error("subject key missing");
    }
    if (!opts.schema) {
        throw new Error("schema key missing");
    }
    // Aggregare the configuration values with defaults
    return util_1.aggregateOptions(optionsDefault, opts);
};
//# sourceMappingURL=config.js.map