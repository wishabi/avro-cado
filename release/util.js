"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.handleError = (err) => {
    if (err.statusCode && err.statusCode === 500) {
        if (err.error &&
            err.error.error_code &&
            (err.error.error_code === 50003 || err.error.error_code === 50002)) {
            return true;
        }
    }
    return false;
};
exports.aggregateOptions = (defaultConf, overrideConf) => {
    const aggOptions = Object.assign({}, defaultConf, overrideConf);
    if (!overrideConf) {
        return aggOptions;
    }
    // determine how to handle unions
    const validWrapOptions = ["always", "never", "auto"];
    if (validWrapOptions.filter(opt => opt === aggOptions.wrapUnions).length === 0) {
        aggOptions.wrapUnions = defaultConf.wrapUnions;
    }
    return aggOptions;
};
//# sourceMappingURL=util.js.map