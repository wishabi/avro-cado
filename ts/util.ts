export const handleError = (err: any): boolean => {
  if (err.statusCode && err.statusCode === 500) {
    if (
      err.error &&
      err.error.error_code &&
      (err.error.error_code === 50003 || err.error.error_code === 50002)
    ) {
      return true;
    }
  }

  return false;
};

export const aggregateOptions = (defaultConf, overrideConf) => {
  const aggOptions = Object.assign({}, defaultConf, overrideConf);

  if (!overrideConf) {
    return aggOptions;
  }

  // determine how to handle unions
  const validWrapOptions: Array<string> = ["always", "never", "auto"];
  if (
    validWrapOptions.filter(opt => opt === aggOptions.wrapUnions).length === 0
  ) {
    aggOptions.wrapUnions = defaultConf.wrapUnions;
  }

  return aggOptions;
};
