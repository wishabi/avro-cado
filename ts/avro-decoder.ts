// import Avro = require("avsc");
import * as Avro from "avsc";
import { processOptions, ACCEPT_HEADERS } from "./config";
import { handleError, aggregateOptions } from "./util";
import { DecoderInfo, DecodeFunc } from "./types/types";
import * as rp from "request-promise";

/**
 * Retrieve the AVRO schema from the schema registry. On a
 * retrieable error, retry up to the configure number of times
 *
 * @param id - the schema id that should be retrieved
 * @return - The AVRO schema requested
 */
const genSchemaRetriever = ({ subject, schemaRegistry, numRetries }) => async (
  id: number
): Promise<any> => {
  const req = {
    uri: `${schemaRegistry}/schemas/ids/${id}`,
    headers: [{ Accept: ACCEPT_HEADERS }],
    json: true,
    resolveWithFullResponse: true
  };

  let schema: any;
  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      schema = JSON.parse((await rp(req)).body.schema);

      // make sure to clear any previous errors
      error = null;
      break;
    } catch (err) {
      // save the error
      error = err;

      if (handleError(err) && i + 1 <= numRetries) {
        // try and try again until we run out of retries
        continue;
      }

      // fatal error
      break;
    }
  }

  if (error) {
    throw new Error(
      `Failed to retrieve schema for subject ${subject} with id ${id} :: ${
        error.message
      }`
    );
  }

  return schema;
};

/**
 * Retrieve the AVRO schema from the schema registry, parse it,
 * and create a resolver to the local AVRO schema.
 *
 * @param subject - The subject under which the schema is registered
 * @param id - the schema id that should be the source for the
 *             schema resolver
 * @param avroSchema - the target AVRO schema for the resolver
 * @return - The AVRO schema resolver used to convert messages
 *           encoded with the specified AVRO schema to the local
 *           AVRO schema format
 */
const genCreateSchemaResolver = ({
  subject,
  retrieveSchema
}: DecoderInfo) => async (
  id: number,
  schema: Avro.Type
): Promise<Avro.Resolver> => {
  /**
   * Retrieve the schema by ID from the schema registry ...
   */
  const msgSchema = await retrieveSchema(id);

  /**
   * ... parse the schema ...
   */
  const avSourceSchema = Avro.Type.forSchema(msgSchema);

  /**
   * ... and create a resolver to the schema we know how to consume
   */

  try {
    return schema.createResolver(avSourceSchema);
  } catch (error) {
    throw new Error(`Incompatible schemas :: ${error.message}`);
  }
};

const genGetSchemaResolver = ({
  subject,
  resolversMap,
  createSchemaResolver
}: DecoderInfo) => async (
  encoded: Buffer | null,
  schema: Avro.Type | null
): Promise<Avro.Resolver> => {
  // Do not attempt to get a resolver
  if (!schema) {
    return null;
  }

  if (!encoded) {
    return null;
  }

  /**
   * Extract the schema id with which the buffer was encoded
   */
  const id: number = encoded.slice(1, 5).readInt32BE(0);

  if (!resolversMap[id]) {
    /**
     * We do not have a resolver from the source schema to the
     * schema we support. Go and get a resolver.
     *
     * IMPORTANT
     *
     * We store a promise of a resolver in our hashmap. The reason is that
     * if we have multipe messages beig handled asynchronously, we do not
     * want all of them to flood the schema registry with REST calls
     * to retrieve the schema
     */
    resolversMap[id] = createSchemaResolver(id, schema);
  }

  return resolversMap[id];
};

const decodePayload = async (
  encoded: Buffer | null,
  schema: Avro.Type | null,
  resolver: Promise<Avro.Resolver>
): Promise<Buffer> => {
  // Do not attempt to avro decode if we have no target schema specified
  if (!schema) {
    return encoded;
  }

  // Do not attempt to avro decode null value
  if (!encoded) {
    return encoded;
  }

  return schema.decode(encoded.slice(5), 0, await resolver).value;
};

const genMessageDecoder = ({
  schema,
  getSchemaResolver
}: DecoderInfo): DecodeFunc => async (buffer: Buffer): Promise<Object> => {
  return await decodePayload(
    buffer,
    schema,
    await getSchemaResolver(buffer, schema)
  );
};

/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/

export const createDecoder = (opts): DecodeFunc => {
  // Aggregare the configuration values with defaults
  const mergedOpts = processOptions(opts);

  // decoder info
  const decoderInfo: DecoderInfo = {
    subject: mergedOpts.subject,
    schema: mergedOpts.schema ? Avro.Type.forSchema(mergedOpts.schema) : null,
    resolversMap: {},
    retrieveSchema: null,
    createSchemaResolver: null,
    getSchemaResolver: null
  };

  decoderInfo.retrieveSchema = genSchemaRetriever(mergedOpts);
  decoderInfo.createSchemaResolver = genCreateSchemaResolver(decoderInfo);
  decoderInfo.getSchemaResolver = genGetSchemaResolver(decoderInfo);

  return genMessageDecoder(decoderInfo);
};
