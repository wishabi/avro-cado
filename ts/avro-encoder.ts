import * as Avro from "avsc";
import axios, { AxiosResponse } from "axios";
import { ACCEPT_HEADERS, processOptions } from "./config";
import { EncodeFunc, Options } from "./types/types";
import { handleError } from "./util";

/**
 * Register the specified schema for the specified topic under the
 * specified subject.
 *
 * On an error eligible for retry, keep trying up to the specified number
 * of times. Once all retries have been exhausted, any error is considered
 * a fatal one.
 *
 *
 * @param subject - the subject under which to register the schema
 * @param schemaRegistry - the schema registry host
 * @param numRetries - the number of retries to register before throwing an error
 * @param schema - the schema to register
 *
 * @return - A Promise holding the id of the schema in the schema registry
 */
export const registerSchema = async ({
  subject,
  schemaRegistry,
  schema,
  numRetries,
}: Options): Promise<number> => {
  // craft the REST call to the schema registry
  const req = {
    method: "POST",
    url: `${schemaRegistry}/subjects/${subject}/versions`,
    headers: { Accept: ACCEPT_HEADERS },
    data: {
      schema: JSON.stringify(schema),
    },
    json: true,
  };

  let error = null;

  // implement retry on certain status/reason codes
  for (let i = 0; i <= numRetries; i += 1) {
    try {
      return ((await axios(req)) as AxiosResponse).data.id;
    } catch (err) {
      // save the error
      error = err;

      if (!handleError(err)) {
        break;
      }
    }
  }

  throw new Error(
    `Failed to register schema for subject ${subject} :: ${error.message}`,
  );
};

/**
 * Create a function that takes a message JSON object
 * and Avro encodes it
 *
 * @param payload - the JSON object to Avro encode
 * @return - The Avro encoded object as a Buffer
 */
export const genMessageEncoder = (
  schema: Avro.Type,
  schemaId: number,
): EncodeFunc => (payload: object): Buffer => {
  if (payload === null) {
    // no payload so return it
    return null;
  }

  // create the avro header
  const header = Buffer.alloc(5);

  // magic byte
  header.writeInt8(0, 0);

  // schema id
  header.writeInt32BE(schemaId, 1);

  // Avro encode the payload
  const buffer: Buffer = schema.toBuffer(payload);

  return Buffer.concat([header, buffer]);
};

/*****************************************************************/
/**                      EXPORTED INTERFACE                     **/
/*****************************************************************/

export const createEncoder = async (opts: Options): Promise<EncodeFunc> => {
  // Aggregate the configuration values with defaults
  const mergedOpts: Options = processOptions(opts);

  return genMessageEncoder(
    Avro.Type.forSchema(mergedOpts.schema, {
      wrapUnions: mergedOpts.wrapUnions,
    }),
    await registerSchema(mergedOpts),
  );
};
