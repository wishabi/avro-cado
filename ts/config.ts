import { aggregateOptions } from "../ts/util";

export const ACCEPT_HEADERS: string =
  "application/vnd.schemaregistry.v1+json, application/vnd.schemaregistry+json, application/json";

const optionsDefault = {
  schemaRegistry: "http://localhost:8081",
  numRetries: 10,
  wrapUnions: "auto"
};

export const processOptions = opts => {
  if (!opts.subject) {
    throw new Error("subject key missing");
  }

  if (!opts.schema) {
    throw new Error("schema key missing");
  }

  // Aggregare the configuration values with defaults
  return aggregateOptions(optionsDefault, opts);
};
