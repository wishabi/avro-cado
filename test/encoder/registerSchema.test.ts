import { registerSchema } from "../../ts/avro-encoder";
import { handleError } from "../../ts/util";
import { ACCEPT_HEADERS } from "../../ts/config";
import * as rp from "request-promise";

const opts = {
  subject: "subject",
  schemaRegistry: "host",
  numRetries: 1,
  schema: "schema"
};

jest.mock("request-promise", () => {
  return jest.fn();
});

describe("registerSchema", () => {
  it("should get the schema ID on success", async () => {
    rp.mockImplementationOnce(params => {
      return {
        body: {
          id: 1
        }
      };
    });

    expect.assertions(3);

    const schemaId = await registerSchema(opts);

    expect(rp).toHaveBeenCalledTimes(1);

    expect(schemaId).toMatchSnapshot();
    expect(rp).toHaveBeenCalledWith({
      body: { schema: '"schema"' },
      headers: { Accept: ACCEPT_HEADERS },
      json: true,
      method: "POST",
      resolveWithFullResponse: true,
      uri: `host/subjects/${opts.subject}/versions`
    });
  });

  /*
     * Retry once and on a retriable error and then fail
     * one all configured retries have been exhausted
     */
  it("should retry 1 times on retriable error", async () => {
    rp
      .mockImplementationOnce(params => {
        throw {
          statusCode: 500,
          error: {
            error_code: 50002,
            message: "500:50002"
          },
          message: "500:50002"
        };
      })
      .mockImplementationOnce(params => {
        throw {
          statusCode: 500,
          error: {
            error_code: 50002,
            message: "500:50002"
          },
          message: "500:50002"
        };
      });

    expect.assertions(2);

    await expect(registerSchema(opts)).rejects.toMatchSnapshot();

    expect(rp).toHaveBeenCalledTimes(3);
  });

  /*
     * Do NOT retry on a NON retriable error even though
     * retries are configured
     */
  it("should retry 0 times on fatal error", async () => {
    rp.mockImplementationOnce(params => {
      throw {
        statusCode: 600,
        error: {
          error_code: 50002,
          message: "600:50002"
        },
        message: "600:50002"
      };
    });

    expect.assertions(2);

    await expect(registerSchema(opts)).rejects.toMatchSnapshot();

    expect(rp).toHaveBeenCalledTimes(4);
  });
});
