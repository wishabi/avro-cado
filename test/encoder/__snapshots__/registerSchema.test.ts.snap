// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerSchema should get the schema ID on success 1`] = `1`;

exports[`registerSchema should retry 0 times on fatal error 1`] = `[Error: Failed to register schema for subject subject :: 600:50002]`;

exports[`registerSchema should retry 1 times on retriable error 1`] = `[Error: Failed to register schema for subject subject :: 500:50002]`;
